File: Breadth-First Search/Topological sort/LET_no207_Course Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        to_next_course = collections.defaultdict(list)
        for a, b in prerequisites:
            indegree[a] += 1
            to_next_course[b].append(a)
            
        queue = collections.deque([i for i in range(numCourses) if indegree[i] == 0])
        while queue:
            cur = queue.popleft()
            numCourses -= 1
            for nxt in to_next_course[cur]:
                indegree[nxt] -= 1
                if indegree[nxt] == 0:
                    queue.append(nxt)
        
        return numCourses == 0

File: Breadth-First Search/Topological sort/LET_no207_Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        unordered_map<int, vector<int>> next_course;
        for (auto& item : prerequisites) {
            indegree[item[0]]++;
            next_course[item[1]].push_back(item[0]);
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0)
                q.push(i);
        }

        while (!q.empty()) {
            int cur = q.front();
            q.pop();
            numCourses--;

            for (int nxt : next_course[cur]) {
                if (--indegree[nxt] == 0)
                    q.push(nxt);
            }
        }

        return numCourses == 0;
    }
};